Schemes cannot be loaded

For whatever reason, additional schemes cannot be loaded. They are saved, but exceptions are thrown when they are loaded when starting the IDE. See the trace below.

```java
com.intellij.diagnostic.PluginException: Cannot init  component state [Plugin: com.fwdekker.randomness]
	at com.intellij.ide.plugins.PluginManagerCore.createPluginException(PluginManagerCore.java:399)
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:12)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:58)
	at com.intellij.diagnostic.PluginException.logPluginError(PluginException.java:75)
	at com.intellij.configurationStore.ComponentStoreImpl.initComponent(ComponentStoreImpl.kt:108)
	at com.intellij.configurationStore.ComponentStoreWithExtraComponents.initComponent(ComponentStoreWithExtraComponents.kt:49)
	at com.intellij.serviceContainer.PlatformComponentManagerImpl.initializeComponent(PlatformComponentManagerImpl.kt:258)
	at com.intellij.serviceContainer.ServiceComponentAdapter.createAndInitialize(ServiceComponentAdapter.kt:56)
	at com.intellij.serviceContainer.ServiceComponentAdapter.access$createAndInitialize(ServiceComponentAdapter.kt:15)
	at com.intellij.serviceContainer.ServiceComponentAdapter$doCreateInstance$$inlined$use$lambda$1.run(ServiceComponentAdapter.kt:44)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:585)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:531)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeNonCancelableSection(CoreProgressManager.java:191)
	at com.intellij.serviceContainer.ServiceComponentAdapter.doCreateInstance(ServiceComponentAdapter.kt:43)
	at com.intellij.serviceContainer.BaseComponentAdapter.getInstanceUncached(BaseComponentAdapter.kt:110)
	at com.intellij.serviceContainer.BaseComponentAdapter.getInstance(BaseComponentAdapter.kt:72)
	at com.intellij.serviceContainer.BaseComponentAdapter.getInstance$default(BaseComponentAdapter.kt:65)
	at com.intellij.serviceContainer.PlatformComponentManagerImpl.getService(PlatformComponentManagerImpl.kt:288)
	at com.intellij.openapi.components.ServiceManager.getService(ServiceManager.java:18)
	at com.fwdekker.randomness.integer.IntegerSettings$Companion.getDefault(IntegerSettings.kt:44)
	at com.fwdekker.randomness.integer.IntegerInsertAction$ArrayAction.<init>(IntegerActions.kt:87)
	at com.fwdekker.randomness.integer.IntegerInsertAction$ArrayAction.<init>(IntegerActions.kt)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.intellij.util.ReflectionUtil.newInstance(ReflectionUtil.java:389)
	at com.intellij.util.ReflectionUtil.newInstance(ReflectionUtil.java:376)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.instantiate(ActionManagerImpl.java:169)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.convertStub(ActionManagerImpl.java:152)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.getActionImpl(ActionManagerImpl.java:508)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.getAction(ActionManagerImpl.java:496)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.preloadActions(ActionManagerImpl.java:1365)
	at com.intellij.openapi.actionSystem.impl.ActionPreloader.preload(ActionPreloader.java:13)
	at com.intellij.openapi.application.Preloader.lambda$null$1(Preloader.java:80)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:163)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:585)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:531)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:150)
	at com.intellij.openapi.application.Preloader.lambda$null$2(Preloader.java:76)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:222)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$200(BoundedTaskExecutor.java:30)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.execute(BoundedTaskExecutor.java:201)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:221)
	at com.intellij.util.concurrency.BoundedTaskExecutor$1.run(BoundedTaskExecutor.java:190)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.intellij.util.xmlb.XmlSerializationException: Cannot deserialize class com.fwdekker.randomness.integer.IntegerSettings
	at com.intellij.configurationStore.JdomSerializerImpl.deserialize(xmlSerializer.kt:96)
	at com.intellij.configurationStore.XmlSerializer.deserialize(xmlSerializer.kt:24)
	at com.intellij.configurationStore.DefaultStateSerializerKt.deserializeState(DefaultStateSerializer.kt:40)
	at com.intellij.configurationStore.StateStorageBase.deserializeState(StateStorageBase.kt:33)
	at com.intellij.configurationStore.StateGetterImpl.getState(StorageBaseEx.kt:57)
	at com.intellij.configurationStore.ComponentStoreImpl.doInitComponent(ComponentStoreImpl.kt:399)
	at com.intellij.configurationStore.ComponentStoreImpl.initComponent(ComponentStoreImpl.kt:363)
	at com.intellij.configurationStore.ComponentStoreImpl.initPersistenceStateComponent(ComponentStoreImpl.kt:130)
	at com.intellij.configurationStore.ComponentStoreImpl.initComponent(ComponentStoreImpl.kt:96)
	... 45 more
Caused by: java.util.NoSuchElementException: List is empty.
	at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:194)
	at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:178)
	at com.intellij.serialization.BaseBeanBindingKt.createUsingKotlin(BaseBeanBinding.kt:49)
	at com.intellij.serialization.BaseBeanBindingKt.access$createUsingKotlin(BaseBeanBinding.kt:1)
	at com.intellij.serialization.BaseBeanBinding.newInstance(BaseBeanBinding.kt:34)
	at com.intellij.serialization.xml.KotlinAwareBeanBinding.newInstance(KotlinAwareBeanBinding.kt:75)
	at com.intellij.util.xmlb.BeanBinding.deserialize(BeanBinding.java:128)
	at com.intellij.util.xmlb.NotNullDeserializeBinding.deserializeUnsafe(NotNullDeserializeBinding.java:14)
	at com.intellij.util.xmlb.Binding.deserializeList(Binding.java:33)
	at com.intellij.util.xmlb.OptionTagBinding.deserialize(OptionTagBinding.java:97)
	at com.intellij.util.xmlb.BasePrimitiveBinding.deserializeUnsafe(BasePrimitiveBinding.java:60)
	at com.intellij.util.xmlb.BeanBinding.deserializeInto(BeanBinding.java:227)
	at com.intellij.util.xmlb.BeanBinding.deserializeInto(BeanBinding.java:181)
	at com.intellij.util.xmlb.BeanBinding.deserialize(BeanBinding.java:129)
	at com.intellij.configurationStore.JdomSerializerImpl.deserialize(xmlSerializer.kt:90)
	... 53 more
```

--- a/.config/detekt/.detekt.yml
+++ b/.config/detekt/.detekt.yml
@@ -1,4 +1,7 @@
 comments:
+    # License in each file is unnecessary.
+    AbsentOrWrongFileLicense:
+        active: false
     # Nothing wrong with documenting private methods.
     CommentOverPrivateFunction:
         active: false

--- a/build.gradle
+++ b/build.gradle
@@ -46,10 +46,10 @@ dependencies {
 /// Configuration
 // Compilation
 compileKotlin {
-    kotlinOptions.jvmTarget = "11"
+    kotlinOptions.jvmTarget = "1.8"
 }
 compileTestKotlin {
-    kotlinOptions.jvmTarget = "11"
+    kotlinOptions.jvmTarget = "1.8"
 }
 
 intellij {

--- a/build.gradle
+++ b/build.gradle
@@ -1,14 +1,14 @@
 /// Plugins
 plugins {
     // Compilation
-    id "org.jetbrains.kotlin.jvm" version "1.3.61"
-    id "org.jetbrains.intellij" version "0.4.16"
+    id "org.jetbrains.kotlin.jvm" version "1.3.72"
+    id "org.jetbrains.intellij" version "0.4.21"
 
     // Tests/coverage
     id "jacoco"
 
     // Static analysis
-    id "io.gitlab.arturbosch.detekt" version "1.4.0"
+    id "io.gitlab.arturbosch.detekt" version "1.9.0"
 
     // Documentation
     id "org.jetbrains.dokka" version "0.9.18"
@@ -46,10 +46,10 @@ dependencies {
 /// Configuration
 // Compilation
 compileKotlin {
-    kotlinOptions.jvmTarget = "1.8"
+    kotlinOptions.jvmTarget = "11"
 }
 compileTestKotlin {
-    kotlinOptions.jvmTarget = "1.8"
+    kotlinOptions.jvmTarget = "11"
 }
 
 intellij {
@@ -97,7 +97,7 @@ dokka {
     outputFormat = "html"
     includes = ["packages.md"]
 
-    jdkVersion = 8
+    jdkVersion = 8 // stdlib documentation is broken in Dokka for JDK 11 (Kotlin/dokka#213)
 
     includeNonPublic = false
     skipDeprecated = false
@@ -108,5 +108,5 @@ dokka {
 // Compatibility checks
 runPluginVerifier {
     pluginFileName = "$rootProject.name-$version"
-    ides = ["IC-2018.1", "IC-2019.1", "IC-2019.3", "CL-2019.3"]
+    ides = ["IC-2018.1", "IC-2019.1", "IC-2020.1", "CL-2019.3"]
 }

--- a/gradle.properties
+++ b/gradle.properties
@@ -5,11 +5,11 @@ kotlin.code.style = official
 
 assertjVersion        = 3.11.1
 assertjSwingVersion   = 3.9.2
-detektVersion         = 1.4.0
+detektVersion         = 1.9.0
 emojiVersion          = 5.1.1
-intellijVersion       = 2019.3
-junitVersion          = 5.6.0
-junitRunnerVersion    = 1.6.0
+intellijVersion       = 2020.1
+junitVersion          = 5.6.2
+junitRunnerVersion    = 1.6.2
 mockitoKotlinVersion  = 2.2.0
 spekVersion           = 2.0.9
-uuidGeneratorVersion  = 3.3.0
+uuidGeneratorVersion  = 4.0.1

--- a/src/main/kotlin/com/fwdekker/randomness/Settings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/Settings.kt
@@ -27,9 +27,11 @@ interface Settings<SELF, SCHEME : Scheme<SCHEME>> : PersistentStateComponent<SEL
      * This field is backed by [currentSchemeName]. If [currentSchemeName] refers to a scheme that is not contained in
      * [schemes], `get`ting this field will throw an exception.
      */
+    @Suppress("UseCheckOrError") // This is shorter and faster
     var currentScheme: SCHEME
         @Transient
-        get() = schemes.first { it.name == currentSchemeName }
+        get() = schemes.firstOrNull { it.name == currentSchemeName }
+            ?: throw IllegalStateException("Current scheme does not exist.")
         set(value) {
             currentSchemeName = value.name
         }

--- a/src/main/kotlin/com/fwdekker/randomness/array/ArraySettings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/array/ArraySettings.kt
@@ -4,9 +4,9 @@ import com.fwdekker.randomness.Scheme
 import com.fwdekker.randomness.Scheme.Companion.DEFAULT_NAME
 import com.fwdekker.randomness.Settings
 import com.fwdekker.randomness.SettingsConfigurable
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.components.State
 import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.service
 import com.intellij.util.xmlb.XmlSerializerUtil
 import com.intellij.util.xmlb.annotations.MapAnnotation
 
@@ -41,7 +41,7 @@ data class ArraySettings(
          * The persistent `ArraySettings` instance.
          */
         val default: ArraySettings
-            get() = ServiceManager.getService(ArraySettings::class.java)
+            get() = service()
     }
 
 

--- a/src/main/kotlin/com/fwdekker/randomness/decimal/DecimalSettings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/decimal/DecimalSettings.kt
@@ -4,9 +4,9 @@ import com.fwdekker.randomness.Scheme
 import com.fwdekker.randomness.Scheme.Companion.DEFAULT_NAME
 import com.fwdekker.randomness.Settings
 import com.fwdekker.randomness.SettingsConfigurable
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.components.State
 import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.service
 import com.intellij.util.xmlb.XmlSerializerUtil
 import com.intellij.util.xmlb.annotations.MapAnnotation
 
@@ -41,7 +41,7 @@ data class DecimalSettings(
          * The persistent `DecimalSettings` instance.
          */
         val default: DecimalSettings
-            get() = ServiceManager.getService(DecimalSettings::class.java)
+            get() = service()
     }
 
 

--- a/src/main/kotlin/com/fwdekker/randomness/integer/IntegerSettings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/integer/IntegerSettings.kt
@@ -4,9 +4,9 @@ import com.fwdekker.randomness.Scheme
 import com.fwdekker.randomness.Scheme.Companion.DEFAULT_NAME
 import com.fwdekker.randomness.Settings
 import com.fwdekker.randomness.SettingsConfigurable
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.components.State
 import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.service
 import com.intellij.util.xmlb.XmlSerializerUtil
 import com.intellij.util.xmlb.annotations.MapAnnotation
 
@@ -44,7 +44,7 @@ data class IntegerSettings(
          * The persistent `IntegerSettings` instance.
          */
         val default: IntegerSettings
-            get() = ServiceManager.getService(IntegerSettings::class.java)
+            get() = service()
     }
 
 

--- a/src/main/kotlin/com/fwdekker/randomness/string/StringSettings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/string/StringSettings.kt
@@ -5,9 +5,9 @@ import com.fwdekker.randomness.Scheme
 import com.fwdekker.randomness.Scheme.Companion.DEFAULT_NAME
 import com.fwdekker.randomness.Settings
 import com.fwdekker.randomness.SettingsConfigurable
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.components.State
 import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.service
 import com.intellij.util.xmlb.XmlSerializerUtil
 import com.intellij.util.xmlb.annotations.MapAnnotation
 import com.intellij.util.xmlb.annotations.Transient
@@ -44,7 +44,7 @@ data class StringSettings(
          * The persistent `StringSettings` instance.
          */
         val default: StringSettings
-            get() = ServiceManager.getService(StringSettings::class.java)
+            get() = service()
     }
 
 

--- a/src/main/kotlin/com/fwdekker/randomness/ui/JBPopupHelper.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/ui/JBPopupHelper.kt
@@ -96,5 +96,6 @@ private fun actionListener(actionPerformed: (ActionEvent?) -> Unit) =
  * @param other the list to multiply with
  * @return the cartesian product of `this` and [other]
  */
+@Suppress("UnusedPrivateMember") // False positive: Used as operator `*`
 private operator fun <E> List<List<E>>.times(other: List<List<E>>) =
     this.flatMap { t1 -> other.map { t2 -> t1 + t2 } }

--- a/src/main/kotlin/com/fwdekker/randomness/uuid/UuidSettings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/uuid/UuidSettings.kt
@@ -5,9 +5,9 @@ import com.fwdekker.randomness.Scheme
 import com.fwdekker.randomness.Scheme.Companion.DEFAULT_NAME
 import com.fwdekker.randomness.Settings
 import com.fwdekker.randomness.SettingsConfigurable
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.components.State
 import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.service
 import com.intellij.util.xmlb.XmlSerializerUtil
 import com.intellij.util.xmlb.annotations.MapAnnotation
 
@@ -42,7 +42,7 @@ data class UuidSettings(
          * The persistent `UuidSettings` instance.
          */
         val default: UuidSettings
-            get() = ServiceManager.getService(UuidSettings::class.java)
+            get() = service()
     }
 
 

--- a/src/main/kotlin/com/fwdekker/randomness/word/WordSettings.kt
+++ b/src/main/kotlin/com/fwdekker/randomness/word/WordSettings.kt
@@ -5,9 +5,9 @@ import com.fwdekker.randomness.Scheme
 import com.fwdekker.randomness.Scheme.Companion.DEFAULT_NAME
 import com.fwdekker.randomness.Settings
 import com.fwdekker.randomness.SettingsConfigurable
-import com.intellij.openapi.components.ServiceManager
 import com.intellij.openapi.components.State
 import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.service
 import com.intellij.util.xmlb.XmlSerializerUtil
 import com.intellij.util.xmlb.annotations.MapAnnotation
 import com.intellij.util.xmlb.annotations.Transient
@@ -43,7 +43,7 @@ data class WordSettings(
          * The persistent `WordSettings` instance.
          */
         val default: WordSettings
-            get() = ServiceManager.getService(WordSettings::class.java)
+            get() = service()
     }
 
 

